amigo-lib/src/main/java/me/ele/amigo/ActivityStub.java

```diff
@@ -0,0 +1,247 @@
+
+    private static final List<Class<? extends ActivityStub>> SINGLE_TOP_STUBS = Arrays.asList(
+            SingleTopStub1.class,
+            SingleTopStub2.class,
+            SingleTopStub3.class,
+            SingleTopStub4.class,
+            SingleTopStub5.class,
+            SingleTopStub6.class,
+            SingleTopStub7.class,
+            SingleTopStub8.class);
+
+    private static final List<Class<? extends ActivityStub>> SINGLE_TASK_STUBS = Arrays.asList(
+            SingleTaskStub1.class,
+            SingleTaskStub2.class,
+            SingleTaskStub3.class,
+            SingleTaskStub4.class,
+            SingleTaskStub5.class,
+            SingleTaskStub6.class,
+            SingleTaskStub7.class,
+            SingleTaskStub8.class);
+
+    private static final List<Class<? extends ActivityStub>> SINGLE_INSTANCE_STUBS = Arrays.asList(
+            SingleInstanceStub1.class,
+            SingleInstanceStub1.class,
+            SingleInstanceStub2.class,
+            SingleInstanceStub3.class,
+            SingleInstanceStub4.class,
+            SingleInstanceStub5.class,
+            SingleInstanceStub6.class,
+            SingleInstanceStub7.class,
+            SingleInstanceStub8.class);
+
+    // [stub activity name : target activity instance]
+    private static final Map<String, ActivityRecord> usedSingleTopStubs = new LinkedHashMap<>(SINGLE_TOP_STUBS.size(), 0.75f, true);
+    private static final Map<String, ActivityRecord> usedSingleTaskStubs = new LinkedHashMap<>(SINGLE_TASK_STUBS.size(), 0.75f, true);
+    private static final Map<String, ActivityRecord> usedSingleInstanceStubs = new LinkedHashMap<>(SINGLE_INSTANCE_STUBS.size(), 0.75f, true);
+
+    private static Class findUnusedActivityStub(String componentName, Map<String, ActivityRecord> recordMap, List<Class<? extends ActivityStub>> stubs) {
+        Log.d("stub", "findUnusedActivityStub: component = " + componentName);
+        for (Class stub : stubs) {
+            ActivityRecord record = recordMap.get(stub.getName());
+            if (record != null && record.activityClazzName.equals(componentName)) {
+                Log.d("stub", "startStubActivity: (reopen) stubClazz = " + stub);
+                return stub;
+            }
+        }
+
+        for (Class stub : stubs) {
+            if (!recordMap.containsKey(stub.getName())) {
+                Log.d("stub", "startStubActivity: stubClazz = " + stub);
+                return stub;
+            }
+        }
+        return null;
+    }
+
+    private static void recycleEarliestSingleTaskStub(Map<String, ActivityRecord> recordMap, int limit) {
+        if (recordMap.size() < limit) {
+            return;
+        }
+
+        Iterator<Map.Entry<String, ActivityRecord>> it = recordMap.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<String, ActivityRecord> entry = it.next();
+            if (entry.getValue().activity != null && !entry.getValue().activity.isFinishing()) {
```

Comment:
I was wrong, I just figure out the activity in `ActivityRecord` will be set in `AmigoInstrumentation#callActivityOnCreate`, before it just something like a stub record.